### PR TITLE
filter2d: fix #13179

### LINKED ISSUE #13179: filter2D crashes when using ROI and kernel size > 11
##### System information (version)
- OpenCV 3.4.3
- Operating System / Platform => Windows 64 Bit
- Compiler => Visual Studio 2017

##### Detailed description

The code below is crashing for kernel sizes > 11, if kernel size is <= 11 it works.

##### Steps to reproduce

```.cpp
    #include <opencv2/opencv.hpp>
    
    int main()
    {
	cv::Mat image = cv::imread("lena.jpg");

	cv::Mat kernel = cv::getGaborKernel(cv::Size(13, 13), 8, 0, 3, 0.25);  // if changed to cv::Size(11, 11), it works

	cv::Mat roi(image, cv::Rect(0, 0, 128, 128));

	cv::Mat filtered(128, 128, roi.type());
	cv::filter2D(roi, filtered, -1, kernel);

	cv::imshow("image", image);
	cv::imshow("filtered", filtered);

	cv::waitKey();

	return 0;
    }
```

### PR DESCRIPTION
**Pull request by take1014**
fix #13179

### CODE REVIEW

File: modules/imgproc/src/filter.cpp [hunk 1]
```diff
@@ -4670,9 +4671,11 @@ static bool dftFilter2D(int stype, int dtype, int kernel_type,
 
     Point anchor = Point(anchor_x, anchor_y);
     Mat kernel = Mat(Size(kernel_width, kernel_height), kernel_type, kernel_data, kernel_step);
+    Mat src = (src_step != dst_step) ? Mat(Size(width, height), stype, src_data, src_step) : Mat(Size(full_width-offset_x, full_height-offset_y), stype, src_data, src_step);
+    Mat dst = (src_step != dst_step) ? Mat(Size(width, height), dtype, dst_data, dst_step) : Mat(Size(full_width, full_height), dtype, dst_data, dst_step);
+    int _width  = (src_step != dst_step) ? width  : full_width;
+    int _height = (src_step != dst_step) ? height : full_height;
```
Comment (alalek) [inline modules/imgproc/src/filter.cpp]: > (src_step != dst_step)

I believe this condition should be improved (steps have no correlation with image sizes / ROIs).

File: modules/imgproc/src/filter.cpp [hunk 2]
```diff
@@ -4671,11 +4672,14 @@ static bool dftFilter2D(int stype, int dtype, int kernel_type,
...
 
-    Mat src(Size(full_width-offset_x, full_height-offset_y), stype, src_data, src_step);
-    Mat dst(Size(full_width, full_height), dtype, dst_data, dst_step);
-    Mat temp;
     int src_channels = CV_MAT_CN(stype);
     int dst_channels = CV_MAT_CN(dtype);
+    Mat src = (src_step / src_channels != dst_step / dst_channels) ? Mat(Size(width, height), stype, src_data, src_step) : Mat(Size(full_width-offset_x, full_height-offset_y), stype, src_data, src_step);
+    Mat dst = (src_step / src_channels != dst_step / dst_channels) ? Mat(Size(width, height), dtype, dst_data, dst_step) : Mat(Size(full_width, full_height), dtype, dst_data, dst_step);
+    int _width  = (src_step / src_channels != dst_step / dst_channels) ? width  : full_width;
+    int _height = (src_step / src_channels != dst_step / dst_channels) ? height : full_height;
```
[the remark this answers is not in the document]
Comment (take1014) [reply modules/imgproc/src/filter.cpp]: Thanks to the follow. I think that "channels" should be changed to "CV_ELEM_SIZE (type)".

[the remark this answers is not in the document]
Comment (alalek) [reply modules/imgproc/src/filter.cpp]: Problem here is with "step" at first. Condition doesn't make any sense (and doesn't work see above).

Probably you should compare `width` vs `full_width` and `offset != 0` to detect ROI (region of interest) case.

[the remark this answers is not in the document]
Comment (take1014) [reply modules/imgproc/src/filter.cpp]: Roi cases are "width != full_width" and "offset_x != 0".

**Comment by vpisarev** [comment]
ok, this looks good as a temporary solution; however, it looks like it solves the problem by disabling DFT-based convolution in some cases, which may result in significant speed degradation when we use ROI of the source image (e.g. we took 10000x10000 image and took 9990x9990 ROI of it to filter using 11x11 kernel - that would be very very slow). Nevertheless, let's merge it and wait for the subsequent, more efficient fixes :+1:

**Comment by take1014** [comment]
@vpisarev Many thanks for your follow-up. I resolved the merge conflict. I will work on improving this problem in the future.

**Pull request merged**